### PR TITLE
Flaky tests

### PR DESCRIPTION
**Pull request by takameyer**
* Looking into flaky tests

## ☑️ ToDos

* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 📝 Update `COMPATIBILITY.md`
* [ ] 🚦 Tests
* [ ] 🔀 Executed flexible sync tests locally if modifying flexible sync
* [ ] 📦 Updated internal package version in consuming `package.json`s (if updating internal packages)
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated

### CODE REVIEW

File: integration-tests/tests/src/tests/sync/user.ts [hunk 1]
```diff
@@ -67,7 +67,11 @@ function removeExistingUsers(): void {
     if (this.app) {
       const users = this.app.allUsers;
       for (const userId of Object.keys(this.app.allUsers)) {
-        await this.app.removeUser(users[userId]);
+        if (users[userId].isLoggedIn) {
+          await this.app.deleteUser(users[userId]);
+        } else {
+          await this.app.removeUser(users[userId]);
+        }
```
Comment (elle-j) [inline integration-tests/tests/src/tests/sync/user.ts:74]: Why was this change needed? From the TS docs, `removeUser()` will handle logging out the user.

File: integration-tests/tests/src/tests/sync/user.ts [hunk 2]
```diff
@@ -102,43 +106,60 @@ describe.skipIf(environment.missingServer, "User", () => {
...
+    // any email address is valid and registerUser() will never fail -- the test doesn't make sense
+    it.skip("invalid email, valid password", async function (this: AppContext & RealmContext) {
+      const credentials = Realm.Credentials.emailPassword({ email: invalidEmail, password: validPassword });
       await expect(this.app.logIn(credentials)).to.be.rejectedWith("invalid username/password"); // this user does not exist yet
       expect(
         this.app.emailPasswordAuth.registerUser({ email: invalidEmail, password: validPassword }),
-      ).to.be.rejectedWith(`failed to confirm user "${invalidEmail}"`);
+      ).to.eventually.be.rejectedWith(`failed to confirm user "${invalidEmail}"`);
       await expect(this.app.logIn(credentials)).to.be.rejectedWith("invalid username/password"); // this user did not register
+    });
```
Comment (elle-j) [inline integration-tests/tests/src/tests/sync/user.ts:137]: Looks like this test belongs in a "custom function" test. I agree with the comment that for an auto confirm app, the login should not be rejected.

**Review by elle-j — APPROVED** [review]
🟢

File: integration-tests/tests/src/tests/sync/user.ts (hunk 1 above)
Comment (takameyer) [reply integration-tests/tests/src/tests/sync/user.ts:74]: This was an attempt to fix the test when I was unsure why a user object was being returned for something that should have failed.  The first assumption was that the user actually did exist.  This can be removed

Comment (elle-j) [reply integration-tests/tests/src/tests/sync/user.ts:74]: Ah okay, thanks for explaining!

Comment (elle-j) [inline integration-tests/tests/src/tests/sync/user.ts:74]: Suggesting this change as per [this thread](https://github.com/realm/realm-js/pull/6032#discussion_r1331456122).
```suggestion
        await this.app.removeUser(users[userId]);
```

File: integration-tests/tests/src/tests/sync/user.ts (hunk 2 above)
Comment (kneth) [reply integration-tests/tests/src/tests/sync/user.ts:133]: We should remove the test instead of skipping it

Comment (elle-j) [reply integration-tests/tests/src/tests/sync/user.ts]: Yeah, you could probably delete that entire `it` test. Additionally, perhaps after the `"custom confirmation function works"` you could add the following (which is pretty much the same test):

```ts
it("custom confirmation function rejects invalid email", async function (this: AppContext & RealmContext) {
      const invalidEmail = randomNonVerifiableEmail();
      const validPassword = "password123456";

      const credentials = Realm.Credentials.emailPassword({ email: invalidEmail, password: validPassword });
      await expect(this.app.logIn(credentials)).to.be.rejectedWith("invalid username/password"); // this user does not exist yet
      expect(
        this.app.emailPasswordAuth.registerUser({ email: invalidEmail, password: validPassword }),
      ).to.eventually.be.rejectedWith(`failed to confirm user "${invalidEmail}"`);
      await expect(this.app.logIn(credentials)).to.be.rejectedWith("invalid username/password"); // this user did not register
    });
```

I know that `describe` suite is currently skipped, but at least we still keep the test around.
